json.go

```diff
@@ -0,0 +1,797 @@
+	if C.dpiOracleTypeNum(node.oracleTypeNum) != C.DPI_ORACLE_TYPE_JSON_ARRAY {
+		return ErrInvalidType
+	}
+	*jsarr = JSONArray{dpiJsonNode: node,
+		dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData)),
+		isMemOwned:   j.isMemOwned}
+	return nil
+}
+
+// Returns JSONScalar from JSON
+func (j JSON) GetJSONScalar(js *JSONScalar, opts JSONOption) error {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	*js = JSONScalar{dpiJsonNode: node,
+		isMemOwned: j.isMemOwned}
+	return nil
+}
+
+// Returns a Go type Value from JSON
+func (j JSON) GetValue(opts JSONOption, val *interface{}) (err error) {
+	var jScalar JSONScalar
+	err = j.GetJSONScalar(&jScalar, opts)
+	if err != nil {
+		if Log != nil {
+			Log("msg", "JSON.GetValue", "Error", err.Error())
+		}
+		return err
+	}
+	*val, err = jScalar.GetValue()
+	if err != nil {
+		if Log != nil {
+			Log("msg", "JSON.GetValue", "Error", err.Error())
+		}
+		return err
+	}
+	return nil
+}
+
+// Returns JSON formatted standard string
+// json library is used, it will be removed
+// with ODPI direct call to get JSON string from JSON object
+// returning empty string for error case, fix?
+func (j JSON) String() string {
+	var jScalar JSONScalar
+	err := j.GetJSONScalar(&jScalar, JSONOptNumberAsString)
+	if err != nil {
+		if Log != nil {
+			Log("msg", "JSON.String", "Error", err.Error())
+		}
+		return ""
+	}
+	jScalarVal, err := jScalar.GetValue()
+	if err != nil {
+		if Log != nil {
+			Log("msg", "JSON.String", "Error", err.Error())
+		}
+		return ""
+	}
+	data, err := json.Marshal(jScalarVal)
+	if err != nil {
+		if Log != nil {
+			Log("msg", "JSON.String", "Error", err.Error())
+		}
+		return ""
+	}
+	return string(data)
+}
+
+func jsonNodeToData(data *Data, node *C.dpiJsonNode) {
+	if node.value == nil {
+		data.dpiData.isNull = 1
+		return
+	}
+	data.dpiData.value = *node.value
+	data.NativeTypeNum = node.nativeTypeNum
+}
+
+// It represents the array input.
+type JSONArray struct {
+	dpiJsonNode  *C.dpiJsonNode
+	dpiJsonArray *C.dpiJsonArray
+	isMemOwned   bool // true indicates associated C memory needs to be freed
+}
+
+func (j JSONArray) getOracleJSONType() C.dpiOracleTypeNum {
+	return C.DPI_ORACLE_TYPE_JSON_ARRAY
+}
+
+func (j JSONArray) Len() int { return int(j.dpiJsonArray.numElements) }
+func (j JSONArray) GetElement(i int) Data {
+	n := int(j.dpiJsonArray.numElements)
+	elts := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonArray.elements)))[:n:n]
+	var d Data
+	jsonNodeToData(&d, &elts[i])
+	return d
+
+}
+
+func (j JSONArray) Get(nodes []Data) []Data {
+	n := int(j.dpiJsonArray.numElements)
+	elts := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonArray.elements)))[:n:n]
+	for i := 0; i < n; i++ {
+		var d Data
+		jsonNodeToData(&d, &elts[i])
+		nodes = append(nodes, d)
+	}
+	return nodes
+}
+
+func getJSONScalarDouble(d Data) (val float64, err error) {
```

Comment:
Where do we really need this? We should convert (lose precision) Oracle NUMBER to float64 only as a last resort.

Reply:
Please see getJSONScalarNumber function, if it looks ok?